### DISCUSSION TITLE
latest git version is broken with nginx proxy

### DISCUSSION THREAD

**Discussion by nipsy** [dani-garcia/vaultwarden#3495] [Q&A]
### Subject of the issue
Updated to the latest version in git (1ec049e2b5f2f1aa3b8d035e94915b131c5e9e25) and everything is broken.  Getting either a 502 bad gateway error or the following error in my nginx logs:

```
[error] 1467253#1467253: *225 upstream prematurely closed connection while reading upstream, client: 1.2.3.4, server: fqdn.com, request: "GET /vault/ HTTP/2.0", upstream: "http://127.0.0.1:8000/vault/", host: "fqdn.com"
```

### Deployment environment

* vaultwarden version: 1ec049e2b5f2f1aa3b8d035e94915b131c5e9e25

* Install method: built from source

* Clients used: web browser

* Reverse proxy and version: nginx 1.22.1-9 via Debian testing/unstable using the exact configuration described in the wiki under "Nginx with sub-path (by BlackDex)"

* MySQL/MariaDB or PostgreSQL version: neither/ SQLite

* Other relevant details:

### Steps to reproduce

Nothing special other than what is described above.  Rolling back to 393645617efa6866001d40cc7f6a9485947e9b23 and  bw_web v2023.3.0b fixes everything again.

### Expected behaviour

Vault page loads in my browser.

### Actual behaviour

502 Bad Gateway or even more cryptic client error due to the stream simply crapping out mid request.

### Troubleshooting data

Nothing much to provide sadly beyond the error above.  I never even see the requests making it through to Rocket as nothing is ever logged there unless I access Rocket directly.  That does at least give me the following:

```
% curl http://127.0.0.1:8000
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>404 Not Found</title>
</head>
<body align="center">
    <div role="main" align="center">
        <h1>404: Not Found</h1>
        <p>The requested resource could not be found.</p>
        <hr />
    </div>
    <div role="contentinfo" align="center">
        <small>Rocket</small>
    </div>
</body>
</html>

% curl http://127.0.0.1:8000/vault/
curl: (52) Empty reply from server
```

When I roll back to the previously working version, that second request actually returns the expected:

```
% curl http://127.0.0.1:8000/vault/
<!doctype html><html class="theme_light"><head><meta charset="utf-8"/><meta name="viewport" content="width=1010"/><meta name="theme-color" content="#175DDC"/><title page-title>Vaultwarden Web Vault</title><link rel="apple-touch-icon" sizes="180x180" href="images/apple-touch-icon.png"/><link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png"/><link rel="icon" type="image/png" sizes="16x16" href="images/favicon-16x16.png"/><link rel="mask-icon" href="images/safari-pinned-tab.svg" color="#175DDC"/><link rel="manifest" href="ca8f66ed7fccfcd0809f.json"/><script defer="defer" src="theme_head.5f24ba8d7aa944e6f52b.js"></script><link href="app/main.c107f7b9354b6317446a.css" rel="stylesheet"></head><body class="layout_frontend"><app-root><div class="mt-5 d-flex justify-content-center"><div><img class="mb-4 logo logo-themed" alt="Bitwarden"/><p class="text-center"><i class="bwi bwi-spinner bwi-spin bwi-2x text-muted" title="Loading" aria-hidden="true"></i></p></div></div></app-root><script defer="defer" src="app/polyfills.428c25638840333a09ee.js"></script><script defer="defer" src="app/vendor.776201439bfdefee32dc.js"></script><script defer="defer" src="app/main.3a27378a743dd4ad9f70.js"></script></body></html>
```

Nothing else has changed, either between the working and broken versions aside from the vaultwarden code and web contents themselves.  The systemd unit file and nginx configurations are identical.  But I can provide whatever else might be helpful.

**Comment by BlackDex** [comment]
Check if you are caching files at the reverse proxy. Try a different browser. Check the logs of nginx.

Make sure you have a trailing `/` at the location for only the sub-path.

**Comment by nipsy** [comment]
Assuming proxy_cache is pretty much the only way to enable caching, there's none of that happening.  Like I said, my configuration mirrors the wiki, aside from also needing to specify 'add_header Content-Security-Policy "";' in each of the three location blocks.

```
upstream vaultwarden-default {
        zone vaultwarden-default 64k;
        server 127.0.0.1:8000;
        keepalive 2;
}

upstream vaultwarden-ws {
        zone vaultwarden-ws 64k;
        server 127.0.0.1:3012;
        keepalive 2;
}
```

```
        location /vault/ {
                add_header Content-Security-Policy "";

                proxy_http_version 1.1;
                proxy_set_header "Connection" "";

                proxy_set_header Host $host;
                proxy_set_header X-Real-IP $remote_addr;
                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                proxy_set_header X-Forwarded-Proto $scheme;

                proxy_pass http://vaultwarden-default;
        }

        location /vault/notifications/hub/negotiate {
                add_header Content-Security-Policy "";

                proxy_http_version 1.1;
                proxy_set_header "Connection" "";

                proxy_set_header Host $host;
                proxy_set_header X-Real-IP $remote_addr;
                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                proxy_set_header X-Forwarded-Proto $scheme;

                proxy_pass http://vaultwarden-default;
        }

        location /vault/notifications/hub {
                add_header Content-Security-Policy "";

                proxy_http_version 1.1;
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection "upgrade";

                proxy_set_header Host $host;
                proxy_set_header X-Real-IP $remote_addr;
                proxy_set_header Forwarded $remote_addr;
                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                proxy_set_header X-Forwarded-Proto $scheme;

                proxy_pass http://vaultwarden-ws;
        }
```

Browser is irrelevant, as curl shows Rocket isn't responding to a simple GET request as expected.

nginx logs themselves aren't very useful beyond the already provided error.  This is immediately after starting the service where I get a 502 Bad Gateway:

```
==> access.log <==
fqdn.com 1.2.3.4 - - [01/May/2023:13:56:18 -0700] "GET /vault/ HTTP/2.0" 502 150 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/112.0"

==> error.log <==
2023/05/01 13:56:18 [error] 1467253#1467253: *258 upstream prematurely closed connection while reading response header from upstream, client: 1.2.3.4, server: fqdn.com, request: "GET /vault/ HTTP/2.0", upstream: "http://127.0.0.1:8000/vault/", host: "bitgnome.net"
```

Hitting reload in my browser at that point gives me a "Problem loading page" in Firefox with "Secure Connection Failed" which initially had me thinking something was wrong with my certificate, but that was a red herring and it's simply because the connection dies unexpectedly.  The logs look pretty similar, albeit without the 502 error this time around:

```
==> access.log <==
fqdn.com 1.2.3.4 - - [01/May/2023:13:58:27 -0700] "GET /vault/ HTTP/2.0" 200 0 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/112.0"

==> error.log <==
2023/05/01 13:58:27 [error] 1467253#1467253: *262 upstream prematurely closed connection while reading upstream, client: 1.2.3.4, server: fqdn.com, request: "GET /vault/ HTTP/2.0", upstream: "http://127.0.0.1:8000/vault/", host: "fqdn.com"
```

In all of that, nothing ever makes it through to Rocket directly anymore, or at least, it's not logging anything:

```
May 01 13:55:52 host.fqdn.com vaultwarden[1513943]: /--------------------------------------------------------------------\
May 01 13:55:52 host.fqdn.com vaultwarden[1513943]: |                        Starting Vaultwarden                        |
May 01 13:55:52 host.fqdn.com vaultwarden[1513943]: |                      Version 1.28.1-1ec049e2                       |
May 01 13:55:52 host.fqdn.com vaultwarden[1513943]: |--------------------------------------------------------------------|
May 01 13:55:52 host.fqdn.com vaultwarden[1513943]: | This is an *unofficial* Bitwarden implementation, DO NOT use the   |
May 01 13:55:52 host.fqdn.com vaultwarden[1513943]: | official channels to report bugs/features, regardless of client.   |
May 01 13:55:52 host.fqdn.com vaultwarden[1513943]: | Send usage/configuration questions or feature requests to:         |
May 01 13:55:52 host.fqdn.com vaultwarden[1513943]: |   https://github.com/dani-garcia/vaultwarden/discussions or        |
May 01 13:55:52 host.fqdn.com vaultwarden[1513943]: |   https://vaultwarden.discourse.group/                             |
May 01 13:55:52 host.fqdn.com vaultwarden[1513943]: | Report suspected bugs/issues in the software itself at:            |
May 01 13:55:52 host.fqdn.com vaultwarden[1513943]: |   https://github.com/dani-garcia/vaultwarden/issues/new            |
May 01 13:55:52 host.fqdn.com vaultwarden[1513943]: \--------------------------------------------------------------------/
May 01 13:55:52 host.fqdn.com vaultwarden[1513943]: [INFO] Using environment file `.env` for configuration.
May 01 13:55:52 host.fqdn.com vaultwarden[1513943]: [2023-05-01 13:55:52.284][vaultwarden::api::notifications][INFO] Starting WebSockets server on 127.0.0.1:3012
May 01 13:55:52 host.fqdn.com vaultwarden[1513943]: [2023-05-01 13:55:52.287][start][INFO] Rocket has launched from http://127.0.0.1:8000
```

Trailing slashes are definitely where they should be.

Like I said, rolling back to the last, known good commit, everything works perfectly.

**Comment by nipsy** [comment]
Oh, I should also mention, I'm running the latest rust:

```
% rustup check
stable-x86_64-unknown-linux-gnu - Up to date : 1.69.0 (84c898d65 2023-04-16)
rustup - Up to date : 1.26.0
```

So unless that's the problem, that shouldn't be causing any issue either.

**Comment by stefan0xC** [comment]
I think this might have been caused by https://github.com/SergioBenitez/Rocket/commit/51ed332127df2737a7d1c1a22a1e13379ea5f6d5

**Reply by tessus** [reply]
I do have to mention that I haven't changed my Apache reverse proxy config and everything still works as it did before. For me nothing breaks.

**Reply by BlackDex** [reply]
I have the exact same nginx config (obviously) and that also still works.

Also, i do not see logs in Vaultwarden, that indicates that nginx can't connect to the server.

**Reply by stefan0xC** [reply]
Do you use subfolders?

**Reply by BlackDex** [reply]
> I think this might have been caused by [SergioBenitez/Rocket@51ed332](https://github.com/SergioBenitez/Rocket/commit/51ed332127df2737a7d1c1a22a1e13379ea5f6d5)

We might need to look into this. And see if we can make sure with and without a trailing `/` it will serve the correct content.

But i think that that also implicaties we need to change something in the web-vault to make sure that also keeps working correctly.

**Reply by BlackDex** [reply]
> Do you use subfolders?

I tested it last weekend with and without sub-path because of an other user reporting an issue. It kept working in all my tests.

**Reply by stefan0xC** [reply]
@BlackDex if I test it locally (without a reverse proxy) it still works but your nginx proxy example seem to have the reported issue. I think this is might be caused by the removed trailing slash from the path. I.e. if using a sub-path there probably should be a `/` if it is significant to rocket (or we use the `AdHoc::uri_normalizer()` to return to the previous behavior)

**Reply by BlackDex** [reply]
I tested via nginx last weekend, and that worked.
I tested both 1.28.1 and my latest commit just merged a few days ago. So i still do not understand what would be different for me.

I'll see if i can test quickly this morning.

**Reply by stefan0xC** [reply]
@BlackDex you've updated the rocket crate which includes this patch only two days ago in f906f6230a8308a3b60d0bac49f99244b7c89df9 so you might not have tested this particular change?

**Reply by BlackDex** [reply]
As said, i tested both 1.28.1 **and** my latest commit which has been merged last weekend via nginx using a sub-path.

So, i don't understand why i didn't see that same issue. Maybe i have some other strange settings, or missed something else 🫤

**Comment by BlackDex** [comment]
I don't know what your last known good version was, but i think it was a bit older.

You need to configure both the following env variables to get it working.
https://github.com/dani-garcia/vaultwarden/blob/1ec049e2b5f2f1aa3b8d035e94915b131c5e9e25/.env.template#L380
https://github.com/dani-garcia/vaultwarden/blob/1ec049e2b5f2f1aa3b8d035e94915b131c5e9e25/.env.template#L72

The values provided there should work for you.

The problem is that you are running on 127.0.0.1.
And nginx reports 502 because it can't connect to the container.

**Comment by nipsy** [comment]
As stated in my original bug report, my last working version was commit 393645617efa6866001d40cc7f6a9485947e9b23 from Apr 24, like a week ago.  I have both the addresses set in my .env:

```
% grep ADDRESS .env
WEBSOCKET_ADDRESS=127.0.0.1
ROCKET_ADDRESS=127.0.0.1
```

and both are reported correctly at startup by Rocket.

Per @stefan0xC it does seem to be a difference in how the trailing slash is getting handled.  This seems to indicate that as a likely problem:

```
% curl http://127.0.0.1:8000/vault/
curl: (18) transfer closed with 47 bytes remaining to read

% curl http://127.0.0.1:8000/vault
<!doctype html><html class="theme_light"><head><meta charset="utf-8"/><meta name="viewport" content="width=1010"/><meta name="theme-color" content="#175DDC"/><title page-title>Vaultwarden Web Vault</title><link rel="apple-touch-icon" sizes="180x180" href="images/apple-touch-icon.png"/><link rel="icon" type="image/png" sizes="32x32" href="images/favicon-32x32.png"/><link rel="icon" type="image/png" sizes="16x16" href="images/favicon-16x16.png"/><link rel="mask-icon" href="images/safari-pinned-tab.svg" color="#175DDC"/><link rel="manifest" href="ca8f66ed7fccfcd0809f.json"/><script defer="defer" src="theme_head.1df11f603fda400762b7.js"></script><link href="app/main.0a5190b263841f0d59ed.css" rel="stylesheet"></head><body class="layout_frontend"><app-root><div class="mt-5 d-flex justify-content-center"><div><img class="mb-4 logo logo-themed" alt="Bitwarden"/><p class="text-center"><i class="bwi bwi-spinner bwi-spin bwi-2x text-muted" title="Loading" aria-hidden="true"></i></p></div></div></app-root><script defer="defer" src="app/polyfills.75c707d99f21eeebfdd8.js"></script><script defer="defer" src="app/vendor.ada4ba3e105977348b84.js"></script><script defer="defer" src="app/main.7bb76e4731bfaa652466.js"></script></body></html>%
```

Whereas on 393645617efa6866001d40cc7f6a9485947e9b23, both with and without the trailing slash are working and returning the expected data.  My .env for that doesn't change between versions:

```
% grep ^DOMAIN .env
DOMAIN=https://fqdn.com/vault/
```

**Reply by BlackDex** [reply]
Strange that i didn't encountered that last weekend.
Then it probably is the commit @stefan0xC mentioned i think.

Not sure why it kept working for me then 🤔

**Reply by tessus** [reply]
My domain is set to:

```
DOMAIN=https://fqdn.com
```

And I can access my vault with `https://fqdn.com` and `https://fqdn.com/`

**Reply by BlackDex** [reply]
What happens if you remove the trailing `/` from the `DOMAIN` variable?

**Reply by nipsy** [reply]
Still broken.  Didn't make the slightest difference unfortunately.

**Comment by BlackDex** [comment]
Ok. I'm not really sure on how to tackle this in a good way. Adding that fairing is a nice workaround maybe.
But I think we should force a redirect to a trailing `/` if that is not the case right now.

Since for the web-vault it does need the trailing `/`, else all the web-vault/static files will be unable to be downloaded.
Something to look into a bit better.

**Reply by stefan0xC** [reply]
I'm not sure why the routes don't have a trailing slash in the first place.
https://github.com/dani-garcia/vaultwarden/blob/1ec049e2b5f2f1aa3b8d035e94915b131c5e9e25/src/main.rs#L523-L524
When do we remove it? Is this a bug in Rocket?

**Reply by BlackDex** [reply]
Looks like the `mount` function removes that for some reason.

**Reply by BlackDex** [reply]
I think we can fix this behaviour by modifying the `web_files` function to this.

```rust
#[get("/<p..>", rank = 10)] // Only match this if the other routes don't match
async fn web_files(p: PathBuf) -> Cached<Option<NamedFile>> {
    // Check if there is no filename given, in that case, we need to redirect to web_index() function.
    if p.file_name().is_none() {
        web_index().await
    } else {
        Cached::long(NamedFile::open(Path::new(&CONFIG.web_vault_folder()).join(p)).await.ok(), true)
    }
}
```

**Reply by BlackDex** [reply]
Hmm, That breaks the `/admin/` too. Not really useful

**Reply by tessus** [reply]
> I'm not sure why the routes don't have a trailing slash in the first place.

This is an old point of discussion. Unfortunately there is no real standard. However, it has always been clear that `/path` and `/path/` are different URIs.

For APIs most developers came to the conclusion that `/path` points to a resource and `/path/` to  a resource group. (But I have also seen different implementations.)
Since there is no standard it's up to the dev how to handle this situtation. If the app (and/or the API spec) allows it, many devs just make them equivalent. I don't know, if Rocket allows to point 2 paths to the same function.

**Reply by stefan0xC** [reply]
@tessus I'm talking about vaultwarden's specific behavior, i.e. we mount the routes to a path like `/test/`  but the routes that are loaded are:

```
[2023-05-02 19:35:58.861][routes][INFO] Routes loaded:
[2023-05-02 19:35:58.861][routes][INFO] GET    /test
[2023-05-02 19:35:58.861][routes][INFO] HEAD   /test
[2023-05-02 19:35:58.862][routes][INFO] GET    /test/<p..> [10]
```

the first two should be `/test/` but they are not and I don't know why.

If I try to reproduce this behavior with [Rockets examples](https://github.com/SergioBenitez/Rocket/tree/master/examples) this does not happen, i.e. they will appear to be correct:

```
Routes:
   >> (index) GET /test/
   >> (upload) POST /test/
   >> (delete) DELETE /test/<id>
   >> (retrieve) GET /test/<id>
   >> (get_file) GET /test/<p..> [10]
```

update: nevermind, this is apparently a bug in Rocket. After deactivating our custom logs and curling the resources I see that our logs were right all along:

```
GET /test/:
   >> Matched: (web_files) GET /test/<p..> [10]
   >> Outcome: Success
   >> Failed to write response: Is a directory (os error 21).
GET /test:
   >> Matched: (web_index) GET /test/
   >> Outcome: Success
   >> Response succeeded.
```

**Reply by BlackDex** [reply]
Thx for the info @tessus .

I think i have found a solution.
That also explains the change for Rocket. Which i think is a good one in this case.

**Reply by tessus** [reply]
@stefan0xC thanks for the explanation. yep, this is indeed interesting.

**Reply by stefan0xC** [reply]
I've create a bug report https://github.com/SergioBenitez/Rocket/issues/2533 because our logs appear to be correct and Rocket's behavior wrong.

**Comment by BlackDex** [comment]
I created a PR which should resolve this: https://github.com/dani-garcia/vaultwarden/pull/3502

**Reply by tessus** [reply]
Will this PR still be necessary after https://github.com/SergioBenitez/Rocket/issues/2533 has been rsolved and a fix is available?

**Reply by stefan0xC** [reply]
@tessus as far as I've tested it it should still be relevant for calling `/admin/` because without this change calling that route (instead of `/admin`) would result in a 404 page.

**Comment by nipsy** [comment]
The latest pull fixed this issue.  Thanks!
